cdap-common/src/main/java/co/cask/cdap/common/io/Locations.java

```diff
@@ -124,6 +128,70 @@ public long size() throws IOException {
   }
 
   /**
+   * Do some processing on the locations contained in the {@code startLocation}, using the {@code processor}. If this
+   * location is a directory, all the locations contained in it will also be processed. If the {@code recursive} tag
+   * is set to true, those locations that are directories will also be processed recursively. If the
+   * {@code startLocation} is not a directory, this method will return the result of the processing of that location.
+   *
+   * @param startLocation location to start the processing from
+   * @param recursive true if this method should be called on the directory {@link Location}s found from
+   *                  {@code startLocation}
+   * @param processor used to process locations. If the {@link Processor#process} method returns false on any
+   *                  {@link Location} object processed, this method will return the current result of the processor.
+   * @param <R> Type of the return value
+   * @throws IOException if the locations could not be read
+   */
+  public static <R> R processLocations(Location startLocation, boolean recursive,
+                                       Processor<LocationStatus, R> processor) throws IOException {
+    boolean process = processor.process(new LocationStatus(startLocation.toURI(), startLocation.length(),
+                                                           startLocation.isDirectory()));
+    if (!process || !startLocation.isDirectory()) {
```

Comment:
You don't need this check if you make it as part of the loop logic:

```java
Deque<FileStatus> stack = Lists.newLinkedList();
stack.push(fs.getFileStatus(new Path(location.toURI())));
while (!stack.isEmpty()) {
  FileStatus status = stack.poll();
  // 1. call processor with status
  // 2. if from status, it's a directory, call list and push to stack
}
```

You can do it in more or less the same for the non HDFS case.